### PR TITLE
Fix datastore grpc doc class generation for python and ruby

### PR DESCRIPTION
**Pull request by swcloud**
#716 This is due to datastore API's GqlQuery message's comment contains '$' so we have to escape it, otherwise it will cause java's regex failure (throwing IAE exception)

In https://github.com/googleapis/googleapis/blob/master/google/datastore/v1/query.proto#L232,

-------------------------

Key must match regex `[A-Za-z_$][A-Za-z_$0-9]*`

-------------------------

The original CommentPatterns.PROTO_LINK_PATTERN causes the above [A-Za-z_$][A-Za-z_$0-9] to be parsed as a proto link, using Ruby as an example (similar with Python and NodeJS)

In RDocCommentFixer.java:
```
  private static String rdocifyProtoMarkdownLinks(String comment) {
    StringBuffer sb = new StringBuffer();
    Matcher m = CommentPatterns.PROTO_LINK_PATTERN.matcher(comment);
    if (!m.find()) {
      return comment;
    }
    do {
      m.appendReplacement(sb, String.format("%s", protoToRubyDoc(m.group(1))));
    } while (m.find());
    m.appendTail(sb);
    return sb.toString();
  }
```

When m.appendReplacement is called, the second argument contains a '$' in it (because m.group(1) is A-Za-z_$, so that causes appendReplacement to throw IAE.

### CODE REVIEW

File: src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java [hunk 1]
```diff
@@ -22,6 +22,8 @@
 
   /** Returns a Sphinx-formatted comment string. */
   public static String sphinxify(String comment) {
+    // escape '$' in the comment first
+    comment = comment.replaceAll("\\$", "\\\\\\$");
```
Comment (bjwatson) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: Why so many backslashes?

Comment (swcloud) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: For each \, you need to two extra \\, any better way?

Comment (jmuk) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: http://docs.oracle.com/javase/7/docs/api/java/lang/String.html#replace(java.lang.CharSequence,%20java.lang.CharSequence) This can be used?

File: src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java [hunk 2]
```diff
@@ -22,6 +22,8 @@
 
   /** Returns a Sphinx-formatted comment string. */
   public static String sphinxify(String comment) {
+    // escape '$' in the comment first
```
Comment (jmuk) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: It is unclear to me why '$' needs to be replaced. Can you bring a URL to explain what '$' sign means in Sphinx?

File: src/main/java/com/google/api/codegen/ruby/RDocCommentFixer.java [hunk 3]
```diff
@@ -21,8 +21,10 @@
 /** Utility class for formatting source comments to follow RDoc style. */
 public class RDocCommentFixer {
 
-  /** Returns a Sphinx-formatted comment string. */
+  /** Returns a RDoc-formatted comment string. */
   public static String rdocify(String comment) {
+    // escape '$' in the comment first
```
Comment (jmuk) [inline src/main/java/com/google/api/codegen/ruby/RDocCommentFixer.java]: ditto for RDoc or YARD link

File: src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java (hunk 1 above)
Comment (geigerj) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: Do we need to undo the extra backslashes at the end so that the generated comment doesn't contain the escape characters?

File: src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java (hunk 2 above)
Comment (swcloud) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: This has nothing to do with Sphinix. This is java regex API's requirement, for example, https://github.com/googleapis/toolkit/blob/master/src/main/java/com/google/api/codegen/ruby/RDocCommentFixer.java#L64, The second argument of 'appendReplacement' cannot contain a non-escaped '$' in it.

File: src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java (hunk 1 above)
Comment (bjwatson) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: 👍 to @jmuk's suggestion. We're not doing any abstract pattern matching here. Literals are fine. (i.e. `fgrep` rather than `egrep`)

Comment (swcloud) [inline src/main/java/com/google/api/codegen/py/PythonSphinxCommentFixer.java]: @jmuk Do you mean we should use 'replace' instead of 'replaceAll'?

File: src/main/java/com/google/api/codegen/CommentPatterns.java [hunk 4]
```diff
@@ -26,6 +26,7 @@
       Pattern.compile("\\[([^\\]]+)\\]\\((\\p{Alpha}+:[^\\)]+)\\)");
   public static final Pattern CLOUD_LINK_PATTERN =
       Pattern.compile("\\[([^\\]]+)\\]\\(((?!\\p{Alpha}+:)[^\\)]+)\\)");
-  public static final Pattern PROTO_LINK_PATTERN = Pattern.compile("\\[([^\\]]+)\\]\\[[^\\]]*\\]");
+  public static final Pattern PROTO_LINK_PATTERN =
+      Pattern.compile("\\[([^\\]]+)\\]\\[([A-Za-z_][A-Za-z_.0-9]*)*\\]");
```
Comment (bjwatson) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: This doesn't look right. I think what you want here is `\\[([^\\]]+)\\]\\[[A-Za-z_][A-Za-z_.0-9]*]*\\]`.

Comment (jmuk) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: I'm not sure why an asterisk appears after `([A-Za-z_][A-Za-z_.0-9]*)`. Why that?

**Comment by bjwatson** [comment]
@jmuk Like fuzzing?

**Comment by jmuk** [comment]
oh, not saying that random :)

I'd like to recommend to copy&paste some comments which causes this problem from the production proto files (or write something by ourselves).

**Comment by bjwatson** [comment]
Got it. SGTM

File: src/main/java/com/google/api/codegen/CommentPatterns.java (hunk 4 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: `?` is better if so. `([A-Za-z_][A-Za-z_.0-9]*)?`

Comment (geigerj) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: Actually, I think I like the one we use [in packman](https://github.com/googleapis/packman/blob/master/lib/api_repo.js#L951) best:`([A-Za-z_]+(\\.[A-Za-z_0-9]+)*)`

Comment (jmuk) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: @geigerj I like that, but I'm afraid that it's not strictly correct, it's going to be `([A-Za-z_][A-Za-z_0-9]*(\\.[A-Za-z_][A-Za-z_0-9]*)*)` in that way.

Comment (swcloud) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: @geigerj But that would not allow the second group to be empty

Comment (swcloud) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: Actually there still be problem with the current approach. For example, [A-Za-z_$][a.b.c.d] is considered to be valid proto link, since we allow '$' within the first pair of [] , so a.b.c.d will be removed. So as long as comment string contains [<anything>][a.b.c.d], it will be parsed as a valid proto link, which will have the problem of removing the content within the second pair of bracket. As @jmuk mentioned, #331 may be the right approach. But in the mean time, we will have to live with is imperfect solution.

**Comment by swcloud** [comment]
@jmuk @bjwatson @geigerj I added unit test cases to test how Python/Ruby/NodeJS CommentFixer should handle '$' to avoid throw IAE. These are not comprehensive test cases for the CommentFixers, because I think this PR should focus on related problems. I filed https://github.com/googleapis/toolkit/issues/722 to add more comprehensive tests. I think @geigerj understands best for these markdown link requirements, so I assigned it to him.

**Comment by swcloud** [comment]
@jmuk I updated the description to add more details how this problem is caused.

File: src/main/java/com/google/api/codegen/CommentPatterns.java (hunk 4 above)
Comment (geigerj) [inline src/main/java/com/google/api/codegen/CommentPatterns.java]: > But that would not allow the second group to be empty such as [book][]

As @jmuk mentioned, you can use the `?` operator to achieve that effect.

> For example, [A-Za-z_$][a.b.c.d] is considered to be valid proto link, since we allow '$' within the first pair of []

I think we should treat `[A-Za-z_$][a.b.c.d]` as a valid proto link for the time being. Without any knowledge of the proto namespace we are working with there is no way to distinguish this programmatically from a link to a proto called `a.b.c.d` with display text `A-Za-z_$`.

**Comment by swcloud** [comment]
@jmuk @bjwatson @geigerj PTAL

File: src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    Matcher m;
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[Shelf][google.example.library.v1.Shelf]");
+    Truth.assertThat(m.find()).isTrue();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[Shelf][]");
+    Truth.assertThat(m.find()).isTrue();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[$Shelf][]");
+    Truth.assertThat(m.find()).isTrue();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[][abc]");
+    Truth.assertThat(m.find()).isFalse();
+    m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[A-Za-z_$][A-Za-z_$0-9]");
```
Comment (bjwatson) [inline src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java]: It's kind of confusing to use something that looks like a regex here. `[$Shelf][$abc]` would fit better with the other examples.

Comment (geigerj) [inline src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java]: It's common for proto comments to contain regular expressions that describe the format of string fields, and those can almost look like markdown links if they have two character classes in a row. (That's what triggered this PR originally...) I think it's OK to retain this as a sort of regression test; Song could also add the example you gave.

Comment (swcloud) [inline src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java]: The reason why [A-Za-z_$][A-Za-z_$0-9] is added as a test case is that it is a corner case causing the issue.

**Comment by bjwatson** [comment]
OK, makes sense. Comments will help make this clear.

**Comment by jmuk** [comment]
I was thinking of adding comments to library.proto (so that you can make sure everything is fine in the baseline files on every language).

It is actually not safe as unit tests you wrote, but it could reveal errors on other languages too.

**Comment by swcloud** [comment]
If we want to comprehensive unit test cases and comments, it should be done in #722 separately.

File: src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java [hunk 6]
```diff
@@ -23,19 +23,25 @@
...
+    // Display name may contain special character '$'
     m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[$Shelf][]");
     Truth.assertThat(m.find()).isTrue();
+    // Display name may NOT be empty
     m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[][abc]");
     Truth.assertThat(m.find()).isFalse();
+    // Fully qualified name may NOT contain special character '$'
     m = CommentPatterns.PROTO_LINK_PATTERN.matcher("[A-Za-z_$][A-Za-z_$0-9]");
     Truth.assertThat(m.find()).isFalse();
+    // Fully qualified name many NOT be numbers only
```
Comment (bjwatson) [inline src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java]: Nit: s/many/may/

Comment (swcloud) [inline src/test/java/com/google/api/codegen/ProtoDocumentLinkTest.java]: Thanks!
